docs/docs/contribute/software/add-new-metric-provider.md

```diff
@@ -0,0 +1,123 @@
+
+1. Fork and clone the [Keptn repo](https://github.com/keptn/lifecycle-toolkit)
+  for more information [checkout this link](https://keptn.sh/stable/docs/contribute/general/git/fork-clone/)
+
+2. **Define the Provider Type:** In the `metrics-operator/controllers/common/providers/common.go` file,
+ define the constant `KeptnPlaceholderProviderType`.
+  In our example we use `"placeholder"`.
+
+    ```go
+    const KeptnPlaceholderProviderType = "placeholder"
+    ```
+
+3. **Implement the Provider:** Create your own new folder inside the
+   [metrics-operator/controllers/common/providers](<https://github.com/keptn/lifecycle-toolkit/tree/main/metrics-operator/controllers/common/providers>),
+   matching the new provider name (`placeholder` in our example).
+   Create a new Go package for the placeholder provider in that folder.
+   This package should contain a `struct` that implements the `KeptnSLIProvider` interface.
+   To fully implement the `KeptnSLIProvider` interface, it's necessary to implement the following functions.
+  
+      * `EvaluateQuery`(Fetches metric values from the provider)
+        * This function fetches metric values based on the provided
+          metric query from the provider.
+          It evaluates the query and returns the metric values
+          along with any additional data if required.
+        * It takes as input a [KeptnMetric](../../reference/crd-reference/metric.md)
+          and [KeptnMetricsProvider](../../reference/crd-reference/metricsprovider.md)
+      * `EvaluateQueryForStep`(Fetches metric values with step interval from the provider)
+        * This function fetches metric values with a specified step interval from the placeholder provider.
+          It takes into account the metric query and the step interval provided, executes the query,
+          and returns the metric values along with any additional data if required.
+        * It takes as input a [KeptnMetric](../../reference/crd-reference/metric.md)
+          and [KeptnMetricsProvider](../../reference/crd-reference/metricsprovider.md)
+      * `FetchAnalysisValue`(Fetches analysis values from the provider) functions.
+        * This function fetches analysis values based on the provided query and time range from the
+          provider.
+          It evaluates the query within the specified time range and returns the analysis
+          values along with any additional data if required.
+        * It takes as input an [Analysis](../../reference/crd-reference/analysis.md),
+          resource that contains a `query` and a
+          [KeptnMetricsProvider](../../reference/crd-reference/metricsprovider.md) resource.
+
+   You can follow other existing implementations,
+   such as [prometheus.go](https://github.com/keptn/lifecycle-toolkit/blob/main/metrics-operator/controllers/common/providers/prometheus/prometheus.go),
+   as an example.
+
+   **NB:** Each of the three functions expects a string containing a float value in it.
```

Comment:
```suggestion
Each of the three functions expects a string containing a float value in it.
```
No need for the "NB".  However, this should be nested under list item #3.  The fact that it isn't is probably why the next step is numbered 1 rather than 4.

I do see the issue -- the code example has some long lines so it is nice to have it start at the left margin.  One thought is to pull the provider implementation and instantiation case out of the list of steps then create level-2 sections -- something like "## Example provider implementation" and ##Example instantiation case" -- and put those after the list of steps.  This also has the advantage of putting them in a page contents list in the right frame so the reader can quickly jump to them.  Then the steps in the list can have the content they currently have and end with a link to the section that has the example.

Reply:
The suggestion didn't work as `markdown-fix` keeps returning level-2 sections to the beginning of the line.

Reply:
Yes, it SHOULD return Level 2 headers to the left margin and reset the automatic numbering.    That is why I suggested putting them in Level 2 sections that are after all the numbered steps.  You can put a link to each example in the step.